executor/src/witgen/jit/solver/mod.rs

```diff
@@ -0,0 +1,212 @@
+use powdr_number::FieldElement;
+use utils::known_variables;
+
+use super::effect::Effect;
+use super::quadratic_symbolic_expression::{Error, RangeConstraintProvider};
+use super::{
+    quadratic_symbolic_expression::QuadraticSymbolicExpression,
+    symbolic_expression::SymbolicExpression,
+};
+use crate::witgen::jit::variable_update::VariableUpdate;
+use crate::witgen::range_constraints::RangeConstraint;
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+
+#[cfg(test)]
+mod tests;
+mod utils;
+
+/// Current state of a variable in the solver.
+#[derive(Debug, PartialEq)]
+struct VariableState<T: FieldElement, V> {
+    /// Range constraint for the variable.
+    range_constraint: RangeConstraint<T>,
+    /// None if unknown, otherwise a symbolic expression that can be used to compute
+    /// the variable from other variables.
+    /// Note that the expression is an instance of `SymbolicExpression::Concrete(_)`,
+    /// we have found a concrete variable assignment.
+    symbolic_expression: Option<SymbolicExpression<T, V>>,
+}
```

Comment:
In this representation, is it the case that 42 can be expressed as 
```
{ range: 42..42, symbolic_expression: None }
{ range: 12..55, symbolic_expression: Some(SymbolicExpression::Concrete(42)) }
```
Do we rely on a concrete value's range being exactly [x, x] by construction?

Reply:
The other issue is that the `SymbolicExpression` will also have a RangeConstraint. How do they relate to each other? Should we instead use an `enum` here?